erts/emulator/beam/erl_bif_lists.c

```diff
@@ -926,6 +970,13 @@ static int subtract_continue(Process *p, ErtsSubtractContext *context) {
                 return res;
             }
 
+            /* If either list is empty then there's nothing to do:
+             * return the left-hand side unchanged. */
+            if (context->lhs_remaining == 0 || context->rhs_remaining == 0) {
+                context->result = context->lhs_original;
```

Comment:
```suggestion
                ASSERT(is_nil(context->lhs_original) && is_nil(context->rhs_original))
                context->result = NIL;
```

Reply:
This branch is reachable when either list is nil actually, so the assertion can fail for `[1, 2, 3] -- []` for example. I split this `if` block into two: one for when the `lhs_remaining == 0` which asserts `lhs_original` is nil and sets the result as `NIL`, and another for `rhs_remaining == 0` which asserts `rhs_original` is nil and sets the result to the `lhs_original`.

Reply:
Err, sorry, brainfart :D